samza-core/src/main/java/org/apache/samza/table/remote/RemoteTableProvider.java

```diff
@@ -0,0 +1,115 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.JavaTableConfig;
+import org.apache.samza.container.SamzaContainerContext;
+import org.apache.samza.table.ReadableTable;
+import org.apache.samza.table.Table;
+import org.apache.samza.table.TableProvider;
+import org.apache.samza.table.TableSpec;
+import org.apache.samza.task.TaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Provide for remote table instances
+ */
+public class RemoteTableProvider implements TableProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteTableProvider.class);
+
+  private final TableSpec tableSpec;
+  private TableReadFunction<?, ?> readFn;
+  private TableWriteFunction<?, ?> writeFn;
+  private SamzaContainerContext containerContext;
+  private TaskContext taskContext;
+
+  public RemoteTableProvider(TableSpec tableSpec) {
+    this.tableSpec = tableSpec;
+  }
+
+  @Override
+  public void init(SamzaContainerContext containerContext, TaskContext taskContext) {
+    this.containerContext = containerContext;
+    this.taskContext = taskContext;
+  }
+
+  @Override
+  public void start() {
+    try {
+      readFn = deserializeObject(tableSpec.getConfig().get(RemoteReadableTable.READ_FN));
+      String strWriter = tableSpec.getConfig().getOrDefault(RemoteReadableTable.WRITE_FN, "");
+      if (!strWriter.isEmpty()) {
+        writeFn = deserializeObject(tableSpec.getConfig().get(RemoteReadableTable.WRITE_FN));
+      }
+      LOG.info("Started.");
+    } catch (Exception e) {
+      String errMsg = "Failed to deserialize " + (readFn == null ? "TableReadFunction" : "TableWriteFunction");
+      throw new SamzaException(errMsg, e);
+    }
+  }
+
+  @Override
+  public void stop() {
+    LOG.info("Stopped.");
+  }
+
+  private <T> T deserializeObject(String strObject) throws IOException, ClassNotFoundException {
+    byte [] bytes = Base64.getDecoder().decode(strObject);
+    return (T) new ObjectInputStream(new ByteArrayInputStream(bytes)).readObject();
+  }
+
+  @Override
+  public Table getTable() {
+    ReadableTable table;
+    if (this.writeFn == null) {
+      table = new RemoteReadableTable(tableSpec.getId(), readFn);
```

Comment:
If the instance of a `ReaderFn` and `WriterFn` is scoped to a table, I'd prefer to have the `ReaderFn` and `WriterFn` objects created here when we instantiate a `new Table`. What do you think?

Reply:
very good point. I'll move the instantiation and init of readFn/writeFn to table provider which is by design a factory-like class. So it makes more sense to do the assembly there.